test/waitforapp_test.js

```diff
@@ -19,47 +50,99 @@ suite('waitforapp', function() {
+      var iframe = client.findElement('iframe[src*="' + domain + '"]');
+
+      assert.ok(iframe);
+      assert.ok(iframe.displayed());
+    });
+  });
+
+  suite('correctly works with inline activity apps', function() {
+    var activityCallerOrigin = 'app://' + ACTIVITY_CALLER_DOMAIN;
+    var activityCalleeOrigin = 'app://' + ACTIVITY_CALLEE_DOMAIN;
+
+    setup(function(done) {
+      launch(client.mozApps, activityCalleeOrigin, done);
+    });
+
+    setup(function(done) {
+      launch(client.mozApps, activityCallerOrigin, done);
+    });
+
+    setup(function(done) {
+      switchToApp(client.mozApps, activityCallerOrigin, function(err) {
+        if (err) {
+          return done(err);
+        }
+
+        // Run activity callee as inline activity
+        client.findElement('#open-activity').tap();
+        client.switchToFrame();
+
         done();
       });
     });
 
-    test('iframe is with the render class', function(done) {
-      client.findElement('iframe[src*="' + domain + '"]', function(err, el) {
-        el.scriptWith(
-          function(el) {
-            return el.parentNode.getAttribute('class');
-          },
-          function(err, value) {
-            assert.ok(value.indexOf('render') !== -1);
-            done();
-          }
+    test('waitForActivity should return activity iframe', function(done) {
+      waitForActivity(client.mozApps, ACTIVITY_CALLEE_DOMAIN,
+      function(err, el) {
+        if (err) {
+          return done(err);
+        }
+
+        assert.ok(el);
+        assert.ok(el.displayed());
+        assert.ok(el.getAttribute('src').indexOf(activityCalleeOrigin) === 0);
+        assert.ok(
+          el.getAttribute('parentapp').indexOf(activityCallerOrigin) === 0
         );
+        assert.ok(getParentClass(el).indexOf('render') >= 0);
+        assert.equal(getTransitionState(el), 'opened');
+
+        done();
       });
     });
 
-    test('the transition-state of the iframe is opened', function(done) {
-      client.findElement('iframe[src*="' + domain + '"]', function(err, el) {
-        el.scriptWith(
-          function(el) {
-            return el.parentNode.parentNode.getAttribute('transition-state');
-          },
-          function(err, value) {
-            assert.equal(value, 'opened');
-            done();
+    test('waitForApp should return non-activity iframe', function(done) {
+      // Focus on main instance
+      launch(client.mozApps, activityCalleeOrigin, function(err) {
+        if (err) {
+          return done(err);
+        }
+
+        waitForApp(client.mozApps, ACTIVITY_CALLEE_DOMAIN, function(err, el) {
+          if (err) {
+            return done(err);
           }
-        );
-      });
-    });
 
-    test('iframe is visible', function(done) {
-      client.findElement('iframe[src*="' + domain + '"]', function(err, el) {
-        el.displayed(function(err, displayed) {
-          assert.ok(displayed);
+          assert.ok(el);
```

Comment:
I joined these asserts into one test case, but can split in several as was done before if you prefer.